slides/advanced-01-tuning-hyperparameters.qmd

```diff
@@ -0,0 +1,696 @@
+
+The data are in the modeldatatoo package. We'll sample down the data and refactor some columns: 
+
+:::: {.columns}
+
+::: {.column width="40%"}
+
+```{r}
+#| label: tune-startup
+library(tidymodels)
+library(modeldatatoo)
+
+# Max's usual settings: 
+tidymodels_prefer()
+theme_set(theme_bw())
+options(
+  pillar.advice = FALSE, 
+  pillar.min_title_chars = Inf
+)
+```
+
+:::
+
+::: {.column width="60%"}
+
+```{r}
+#| label: data-import
+set.seed(295)
+hotel_rates <- 
+  data_hotel_rates() %>% 
+  sample_n(5000) %>% 
+  arrange(arrival_date) %>% 
+  select(-arrival_date_num, -arrival_date) %>% 
+  mutate(
+    company = factor(as.character(company)),
+    country = factor(as.character(country)),
+    agent = factor(as.character(agent))
+  )
+```
+
+
+:::
+
+::::
+
+
+## Data Spending
+
+Let's split the data into a training set (75%) and testing set (25%):
+
+```{r}
+#| label: hotel-split
+set.seed(4028)
+hotel_split <-
+  initial_split(hotel_rates, strata = avg_price_per_room)
+
+hotel_tr <- training(hotel_split)
+hotel_te <- testing(hotel_split)
+```
+
+
+
+## Your turn {transition="slide-in"}
+
+Let's take some time and investigate the _training data_. The outcome is `avg_price_per_room`. 
+
+Are there any characteristics of the data that are interesting?
+
+```{r}
+#| label: hotel-investigation-exercise
+#| echo: false
+countdown(minutes = 10, id = "hotel-investigation")
+```
+
+
+## Feature Engineering
+
+How should we represent our predictors? There are: 
+
+ * Categorical predictors, some with many levels
+ * Some skewed numeric predictor(s)
+ * Interactions? 
+
+. . .
+
+Let's start a basic recipe: 
+
+```{r}
+#| label: hotel-rec-start
+
+hotel_rec <- 
+  recipe(avg_price_per_room ~ ., data = hotel_tr) %>% 
+  step_YeoJohnson(lead_time)
+```
+
+. . .
+
+We'll add to this for different types of models. 
+
+## A Baseline Model
+
+A simple linear model with main effects
+
+```{r}
+#| label: hotel-base-model
+set.seed(472)
+hotel_rs <- vfold_cv(hotel_tr, strata = avg_price_per_room)
+
+base_rec <- 
+  hotel_rec %>% 
+  step_dummy(all_nominal_predictors()) %>% 
+  step_zv(all_predictors())
+
+base_wflow <- workflow(base_rec, linear_reg())
+```
+
+
+## A Baseline Model
+
+We'll try to optimize the mean absolute error: 
+
+```{r}
+#| label: hotel-base-res
+#| cache: true
+reg_metrics <- metric_set(mae, rsq)
+ctrl_rs <- control_resamples(save_pred = TRUE)
+
+base_res <- base_wflow %>% fit_resamples(hotel_rs, metrics = reg_metrics)
+
+collect_metrics(base_res)
+```
+
+
+# Optimizing Models via Tuning Parameters
+
+
+
+## Tuning parameters
+
+Some model or preprocessing parameters cannot be estimated directly from the data.
+
+. . .
+
+Some examples:
+
+- Tree depth in decision trees
+- Number of neighbors in a K-nearest neighbor model
```

Comment:
Since we mention that preprocessing parameters are tunable, we should list an example